File: backend/controllers/travel.js
const mongoose = require('mongoose');

const travelModel = require('../models/travel');
const busModel = require('../models/bus');
const stopModel = require('../models/stop');


exports.getTravelBuses = async (req, res) => {
    try {
        const { source, destination, date } = req.query;

        console.log(date);

        if (!source || !destination || !date) {
            return res.status(400).json({
                message: "Source, Destination and Date are required",
                success: false
            })
        }

        if (source === destination) {
            return res.status(400).json({
                message: "Source and Destination can't be same",
                success: false
            })
        }

        const sourceStop = await stopModel.findOne({ city: source });
        const destinationStop = await stopModel.findOne({ city: destination });

        if (!sourceStop || !destinationStop) {
            return res.status(404).json({
                message: "Source or Destination not found",
                success: false
            })
        }

        const travelsBuses = await travelModel.find({
            $or:[
                {departure: {$gte: new Date(date)}},
            ]
        });
        console.log(travelsBuses)
        let finalBuses = [];
        for (let travel of travelsBuses) {
            const bus = await busModel.findById(travel.busId).populate('stops').exec();
            console.log(bus, 'kksdjflkajslkdf')
            for (let i = 0; i < bus.stops.length; i++) {
                if (bus.stops[i].city.toString() === source) {
                    for (let j = i + 1; j < bus.stops.length; j++) {
                        if (bus.stops[j].city.toString() === destination) {
                            finalBuses.push(travel);
                        }
                    }
                }
            }
        }

        if (finalBuses.length === 0) {
            return res.status(404).json({
                message: "No buses found",
                success: false
            })
        }

        return res.status(200).json({
            message: "Travel buses found",
            success: true,
            data: finalBuses
        })
    } catch (error) {
        console.log(error)
        return res.status(500).json({
            message: "Something went wrong in getting travel buses",
            success: false,
            error: error.message
        })
    }
}

exports.createTravel = async (req, res) => {
    try {
        const { busId,
            source,
            destination,
            departure,
            arrival,
            price,
            seatCapacity,
            availableSeats,
            distance } = req.body;
        console.log("hello world 143")
        console.log(departure)
        if (!busId || !source || !destination || !departure || !arrival || !price || !seatCapacity || !availableSeats || !distance) {
            return res.status(400).json({
                message: "All fields are required",
                success: false
            })
        }

        if (!mongoose.Types.ObjectId.isValid(busId)) {
            return res.status(400).json({
                message: "Invalid bus id",
                success: false
            })
        }

        const bus = await busModel.findById(busId);

        if (!bus) {
            return res.status(404).json({
                message: "Bus not found",
                success: false
            })
        }

        const sourceStop = await stopModel.findById(source);
        const destinationStop = await stopModel.findById(destination)

        if (!sourceStop || !destinationStop) {
            return res.status(404).json({
                message: "Source or Destination not found",
                success: false
            })
        }

        const travel = new travelModel({
            busId,
            source: sourceStop._id,
            destination: destinationStop._id,
            departure: Date.parse(departure),
            arrival: Date.parse(arrival),
            price,
            seatCapacity,
            availableSeats,
            distance
        })

        await travel.save();

        return res.status(201).json({
            message: "Travel created successfully",
            success: true,
            data: travel
        })
    } catch (error) {
        return res.status(500).json({
            message: "Something went wrong in creating travel",
            success: false,
            error: error.message
        })
    }
}

File: app/lib/app/models/travel.dart
import 'package:app/app/models/stop.dart';

class Travel {
  String? sId;
  BusId? busId;
  Stop? source;
  Stop? destination;
  DateTime? departure;
  DateTime? arrival;
  int? seatCapacity;
  int? price;
  int? availableSeats;
  int? distance;

  Travel(
      {this.sId,
      this.busId,
      this.source,
      this.destination,
      this.departure,
      this.seatCapacity,
      this.arrival,
      this.price,
      this.availableSeats,
      this.distance});

  Travel.fromJson(Map<String, dynamic> json) {
    sId = json['_id'];
    busId = json['busId'] != null ? BusId.fromJson(json['busId']) : null;
    source =
        json['source'] != null ?  Stop.fromJson(json['source']) : null;
    destination = json['destination'] != null
        ?  Stop.fromJson(json['destination'])
        : null;
    departure = DateTime.parse(json['departure']);
    arrival = DateTime.parse(json['arrival']);
    seatCapacity = json['seatCapacity'];
    price = json['price'];
    availableSeats = json['availableSeats'];
    distance = json['distance'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data =  Map<String, dynamic>();
    data['_id'] = sId;
    if (busId != null) {
      data['busId'] = busId!.toJson();
    }
    if (source != null) {
      data['source'] = source!.toJson();
    }
    if (destination != null) {
      data['destination'] = destination!.toJson();
    }
    data['departure'] = departure;
    data['seatCapacity'] = seatCapacity;
    data['arrival'] = arrival;
    data['price'] = price;
    data['availableSeats'] = availableSeats;
    data['distance'] = distance;
    return data;
  }
}

class BusId {
  String? sId;
  String? name;
  String? number;
  int? seatCapacity;
  String? sourceStop;
  String? destinationStop;
  List<String>? stops;
  String? parkingAddress;
  String? ownerId;
  String? busDetails;

  BusId(
      {this.sId,
      this.name,
      this.number,
      this.seatCapacity,
      this.sourceStop,
      this.destinationStop,
      this.stops,
      this.parkingAddress,
      this.ownerId,
      this.busDetails});

  BusId.fromJson(Map<String, dynamic> json) {
    sId = json['_id'];
    name = json['name'];
    number = json['number'];
    seatCapacity = json['seatCapacity'];
    sourceStop = json['sourceStop'];
    destinationStop = json['destinationStop'];
    stops = json['stops'].cast<String>();
    parkingAddress = json['parkingAddress'];
    ownerId = json['ownerId'];
    busDetails = json['busDetails'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data =  Map<String, dynamic>();
    data['_id'] = sId;
    data['name'] = name;
    data['number'] = number;
    data['seatCapacity'] = seatCapacity;
    data['sourceStop'] = sourceStop;
    data['destinationStop'] = destinationStop;
    data['stops'] = stops;
    data['parkingAddress'] = parkingAddress;
    data['ownerId'] = ownerId;
    data['busDetails'] = busDetails;
    return data;
  }
}

